File: packages/lhc/src/turns/internal/chunks.ts
// Chunk mechanics: open-chunk append, the accumulated close policy, and the
// close-to-summary enqueues. Placement is pure arithmetic over stored projected
// token counts. The caller hands in the incoming turn's count, the open chunk
// carries its accumulated count durably, and placement never uses clock,
// inference, or token re-estimation. Everything here runs inside the
// turn-derivation completion transaction; a crash leaves either a placed turn
// with any close's summary enqueues or nothing.
import type { DatabaseSync } from "node:sqlite";
import type { DbWriteTransaction } from "../../shared-tech/index.js";
import { enqueue, supersedeQueued, type WorkItemRecord } from "../../shared-tech/work-queue/index.js";

export interface ChunkPolicy {
  targetProjectedTokens: number;
  maxProjectedTokens: number;
}

export interface PlacementResult {
  chunkId: string;
  memberIdx: number;
  // Chunks closed by this placement, in close order: at most the previous
  // open chunk (accumulation rule) and then the incoming turn's own chunk
  // (max rule) — a turn ≥ max arriving behind an open chunk closes both.
  closedChunkIds: string[];
  // An explicitly rebuilt turn keeps its placement. Sanctioned delete rebuilds
  // chunk summaries from surviving members without changing chunk membership.
  alreadyPlaced: boolean;
}

interface OpenChunkRow {
  chunk_id: string;
  accumulated_projected_tokens: number | bigint;
}

function memberCount(db: DatabaseSync, chunkId: string): number {
  const row = db.prepare(`SELECT COUNT(*) AS n FROM chunk_member WHERE chunk_id = ?`).get(chunkId) as unknown as {
    n: number | bigint;
  };
  return Number(row.n);
}

function closeChunk(db: DatabaseSync, chunkId: string): void {
  db.prepare(`UPDATE chunk SET status = 'closed' WHERE chunk_id = ?`).run(chunkId);
}

function openNextChunk(db: DatabaseSync): string {
  const row = db.prepare(`SELECT MAX(chunk_order) AS max_order FROM chunk`).get() as
    | { max_order: number | bigint | null }
    | undefined;
  const order = Number(row?.max_order ?? 0) + 1;
  const chunkId = `c${order}`;
  db.prepare(
    `INSERT INTO chunk (chunk_id, chunk_order, status, accumulated_projected_tokens)
     VALUES (?, ?, 'open', 0)`,
  ).run(chunkId, order);
  return chunkId;
}

// The close policy: when the open chunk's accumulated count plus the incoming
// turn's count would reach the target, equality included, the chunk closes
// holding its current members and the incoming turn opens the next chunk. A
// single turn at or above max closes its own chunk immediately. An empty open
// chunk always accepts the incoming turn; a chunk never closes empty.
export function placeTurn(
  db: DatabaseSync,
  turnId: string,
  projectedTokens: number,
  policy: ChunkPolicy,
): PlacementResult {
  const existing = db
    .prepare(`SELECT chunk_id, member_idx FROM chunk_member WHERE turn_id = ?`)
    .get(turnId) as unknown as { chunk_id: string; member_idx: number | bigint } | undefined;
  if (existing !== undefined) {
    return {
      chunkId: existing.chunk_id,
      memberIdx: Number(existing.member_idx),
      closedChunkIds: [],
      alreadyPlaced: true,
    };
  }

  const closedChunkIds: string[] = [];
  let open = db
    .prepare(
      `SELECT chunk_id, accumulated_projected_tokens FROM chunk
       WHERE status = 'open' ORDER BY chunk_order DESC LIMIT 1`,
    )
    .get() as unknown as OpenChunkRow | undefined;

  if (
    open !== undefined &&
    memberCount(db, open.chunk_id) > 0 &&
    Number(open.accumulated_projected_tokens) + projectedTokens >= policy.targetProjectedTokens
  ) {
    // Crossing closes without the incoming turn: the decision weighs
    // accumulated + incoming, and the incoming turn starts the next chunk.
    closeChunk(db, open.chunk_id);
    closedChunkIds.push(open.chunk_id);
    open = undefined;
  }

  const chunkId = open?.chunk_id ?? openNextChunk(db);
  const memberIdx = memberCount(db, chunkId);
  db.prepare(`INSERT INTO chunk_member (chunk_id, turn_id, member_idx) VALUES (?, ?, ?)`).run(
    chunkId,
    turnId,
    memberIdx,
  );
  db.prepare(
    `UPDATE chunk SET accumulated_projected_tokens = accumulated_projected_tokens + ?
     WHERE chunk_id = ?`,
  ).run(projectedTokens, chunkId);

  if (projectedTokens >= policy.maxProjectedTokens) {
    // Max rule: the incoming turn alone meets the maximum, so its chunk closes
    // immediately, whatever it holds.
    closeChunk(db, chunkId);
    closedChunkIds.push(chunkId);
  }

  return { chunkId, memberIdx, closedChunkIds, alreadyPlaced: false };
}

// Closing queues the two summary kinds as independent work items. Both enqueues
// ride the caller's ambient completion transaction.
export function enqueueChunkSummaries(transaction: DbWriteTransaction, chunkId: string): WorkItemRecord[] {
  return (["chunk_summary_detailed", "chunk_summary_brief"] as const).map((kind) =>
    enqueue(transaction, {
      owner: "turns",
      kind,
      sourceRef: { chunkId },
      derivations: [{ subjectKind: "chunk", subjectId: chunkId, derivationType: kind }],
    }),
  );
}

// The chunk structure for compact selection: every chunk in chunk order with
// its raw membership in member order. Membership is NOT filtered by turn
// liveness — references are returned as stored so the consumer can run its
// own referential check (a member pointing at no turn row at all is damage;
// a member pointing at a tombstoned turn is not). This is why it does not
// reuse readChunkRows, whose live-turn join would hide both cases.
export interface ChunkStructureRow {
  chunkId: string;
  chunkOrder: number;
  status: "open" | "closed";
  memberTurnIds: string[];
}

export function readChunkStructure(db: DatabaseSync): ChunkStructureRow[] {
  const chunkRows = db
    .prepare(`SELECT chunk_id, chunk_order, status FROM chunk ORDER BY chunk_order`)
    .all() as unknown as Array<{ chunk_id: string; chunk_order: number | bigint; status: string }>;
  const memberRows = db
    .prepare(
      `SELECT cm.chunk_id, cm.turn_id FROM chunk_member cm
       JOIN chunk c ON c.chunk_id = cm.chunk_id
       ORDER BY c.chunk_order, cm.member_idx`,
    )
    .all() as unknown as Array<{ chunk_id: string; turn_id: string }>;
  const membersByChunk = new Map<string, string[]>();
  for (const row of memberRows) {
    const members = membersByChunk.get(row.chunk_id) ?? [];
    members.push(row.turn_id);
    membersByChunk.set(row.chunk_id, members);
  }
  return chunkRows.map((row) => ({
    chunkId: row.chunk_id,
    chunkOrder: Number(row.chunk_order),
    status: row.status as "open" | "closed",
    memberTurnIds: membersByChunk.get(row.chunk_id) ?? [],
  }));
}

// A tombstoned turn remains a legitimate stable-address record, but a derived
// chunk whose every member turn is tombstoned has no readable source. Such a
// chunk must not survive as an input to compact: its summaries describe content
// the readable projection intentionally removed. Cleanup is deliberately
// revalidated at write time because preview/selection may have read the file
// before another process acquired the compact transaction.
export function dropEmptyReadableChunks(db: DatabaseSync, candidates: readonly string[]): string[] {
  const dropped: string[] = [];
  const hasReadableMember = db.prepare(
    `SELECT 1 FROM chunk_member cm
     JOIN turns t ON t.turn_id = cm.turn_id AND t.deleted_at IS NULL
     WHERE cm.chunk_id = ? LIMIT 1`,
  );
  const hasClaimedWork = db.prepare(
    `SELECT 1 FROM work_item
     WHERE status = 'claimed' AND source_ref = ?
       AND kind IN ('chunk_summary_detailed', 'chunk_summary_brief')
     LIMIT 1`,
  );
  const dropDerivations = db.prepare(`DELETE FROM derivation WHERE subject_kind = 'chunk' AND subject_id = ?`);
  const dropMembers = db.prepare(`DELETE FROM chunk_member WHERE chunk_id = ?`);
  const dropChunk = db.prepare(`DELETE FROM chunk WHERE chunk_id = ?`);

  for (const chunkId of [...new Set(candidates)]) {
    if (hasReadableMember.get(chunkId) !== undefined) continue;
    const sourceRef = { chunkId };
    if (hasClaimedWork.get(JSON.stringify(sourceRef)) !== undefined) continue;
    supersedeQueued(db, [
      { kind: "chunk_summary_detailed", sourceRef },
      { kind: "chunk_summary_brief", sourceRef },
    ]);
    dropDerivations.run(chunkId);
    dropMembers.run(chunkId);
    dropChunk.run(chunkId);
    dropped.push(chunkId);
  }
  return dropped;
}

// Placement read-back for the turns surface: chunkId + memberIdx by turn, one
// query, stored values only.
export function readPlacements(db: DatabaseSync): Map<string, { chunkId: string; memberIdx: number }> {
  const rows = db.prepare(`SELECT turn_id, chunk_id, member_idx FROM chunk_member`).all() as unknown as Array<{
    turn_id: string;
    chunk_id: string;
    member_idx: number | bigint;
  }>;
  const byTurn = new Map<string, { chunkId: string; memberIdx: number }>();
  for (const row of rows) {
    byTurn.set(row.turn_id, { chunkId: row.chunk_id, memberIdx: Number(row.member_idx) });
  }
  return byTurn;
}

File: packages/lhc-py/src/lhc/turns/internal/chunks.py
"""Ported from packages/lhc/src/turns/internal/chunks.ts.

Chunk mechanics: open-chunk append, the accumulated close policy, and the
close-to-summary enqueues. Placement is pure arithmetic over stored projected
token counts. The caller hands in the incoming turn's count, the open chunk
carries its accumulated count durably, and placement never uses clock,
inference, or token re-estimation. Everything here runs inside the
turn-derivation completion transaction; a crash leaves either a placed turn
with any close's summary enqueues or nothing.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

from ...shared_tech.storage import Database
from ...shared_tech.work_queue import WorkItemRecord

if TYPE_CHECKING:
    from ...shared_tech.persist import DbWriteTransaction

_SQL_MEMBER_COUNT = """SELECT COUNT(*) AS n FROM chunk_member WHERE chunk_id = ?"""

_SQL_CLOSE_CHUNK = """UPDATE chunk SET status = 'closed' WHERE chunk_id = ?"""

_SQL_MAX_CHUNK_ORDER = """SELECT MAX(chunk_order) AS max_order FROM chunk"""

_SQL_INSERT_OPEN_CHUNK = (
    """INSERT INTO chunk (chunk_id, chunk_order, status, accumulated_projected_tokens)
     VALUES (?, ?, 'open', 0)"""
)

_SQL_SELECT_EXISTING_PLACEMENT = """SELECT chunk_id, member_idx FROM chunk_member WHERE turn_id = ?"""

_SQL_SELECT_OPEN_CHUNK = (
    """SELECT chunk_id, accumulated_projected_tokens FROM chunk
       WHERE status = 'open' ORDER BY chunk_order DESC LIMIT 1"""
)

_SQL_INSERT_CHUNK_MEMBER = """INSERT INTO chunk_member (chunk_id, turn_id, member_idx) VALUES (?, ?, ?)"""

_SQL_ACCUMULATE_PROJECTED_TOKENS = (
    """UPDATE chunk SET accumulated_projected_tokens = accumulated_projected_tokens + ?
     WHERE chunk_id = ?"""
)

_SQL_SELECT_CHUNK_STRUCTURE = """SELECT chunk_id, chunk_order, status FROM chunk ORDER BY chunk_order"""

_SQL_SELECT_CHUNK_STRUCTURE_MEMBERS = (
    """SELECT cm.chunk_id, cm.turn_id FROM chunk_member cm
       JOIN chunk c ON c.chunk_id = cm.chunk_id
       ORDER BY c.chunk_order, cm.member_idx"""
)

_SQL_SELECT_PLACEMENTS = """SELECT turn_id, chunk_id, member_idx FROM chunk_member"""


@dataclass(frozen=True, slots=True)
class ChunkPolicy:
    target_projected_tokens: int
    max_projected_tokens: int


@dataclass(frozen=True, slots=True)
class PlacementResult:
    chunk_id: str
    member_idx: int
    # Chunks closed by this placement, in close order: at most the previous
    # open chunk (accumulation rule) and then the incoming turn's own chunk
    # (max rule) — a turn ≥ max arriving behind an open chunk closes both.
    closed_chunk_ids: list[str]
    # An explicitly rebuilt turn keeps its placement. Sanctioned delete rebuilds
    # chunk summaries from surviving members without changing chunk membership.
    already_placed: bool


@dataclass(frozen=True, slots=True)
class _OpenChunkRow:
    chunk_id: str
    accumulated_projected_tokens: int


def _member_count(db: Database, chunk_id: str) -> int:
    row = db.prepare(_SQL_MEMBER_COUNT).get(chunk_id)
    return int(row["n"]) if row is not None else 0


def _close_chunk(db: Database, chunk_id: str) -> None:
    db.prepare(_SQL_CLOSE_CHUNK).run(chunk_id)


def _open_next_chunk(db: Database) -> str:
    row = db.prepare(_SQL_MAX_CHUNK_ORDER).get()
    value = row["max_order"] if row is not None else None
    order = int(value if value is not None else 0) + 1
    chunk_id = f"c{order}"
    db.prepare(_SQL_INSERT_OPEN_CHUNK).run(chunk_id, order)
    return chunk_id


# The close policy: when the open chunk's accumulated count plus the incoming
# turn's count would reach the target, equality included, the chunk closes
# holding its current members and the incoming turn opens the next chunk. A
# single turn at or above max closes its own chunk immediately. An empty open
# chunk always accepts the incoming turn; a chunk never closes empty.
def place_turn(
    db: Database,
    turn_id: str,
    projected_tokens: int,
    policy: ChunkPolicy,
) -> PlacementResult:
    existing = db.prepare(_SQL_SELECT_EXISTING_PLACEMENT).get(turn_id)
    if existing is not None:
        return PlacementResult(
            chunk_id=str(existing["chunk_id"]),
            member_idx=int(existing["member_idx"]),
            closed_chunk_ids=[],
            already_placed=True,
        )
    closed: list[str] = []
    row = db.prepare(_SQL_SELECT_OPEN_CHUNK).get()
    open_chunk = (
        _OpenChunkRow(
            chunk_id=str(row["chunk_id"]),
            accumulated_projected_tokens=int(row["accumulated_projected_tokens"]),
        )
        if row is not None
        else None
    )
    target = getattr(policy, "target_projected_tokens")
    maximum = getattr(policy, "max_projected_tokens")
    if (
        open_chunk is not None
        and _member_count(db, open_chunk.chunk_id) > 0
        and open_chunk.accumulated_projected_tokens + projected_tokens >= target
    ):
        _close_chunk(db, open_chunk.chunk_id)
        closed.append(open_chunk.chunk_id)
        open_chunk = None
    chunk_id = open_chunk.chunk_id if open_chunk is not None else _open_next_chunk(db)
    member_idx = _member_count(db, chunk_id)
    db.prepare(_SQL_INSERT_CHUNK_MEMBER).run(chunk_id, turn_id, member_idx)
    db.prepare(_SQL_ACCUMULATE_PROJECTED_TOKENS).run(projected_tokens, chunk_id)
    if projected_tokens >= maximum:
        _close_chunk(db, chunk_id)
        closed.append(chunk_id)
    return PlacementResult(chunk_id, member_idx, closed, False)


# Closing queues the two summary kinds as independent work items. Both enqueues
# ride the caller's ambient completion transaction.
def enqueue_chunk_summaries(transaction: DbWriteTransaction, chunk_id: str) -> list[WorkItemRecord]:
    from ...shared_tech.work_queue import EnqueueDerivationTarget, EnqueueInput, enqueue

    return [
        enqueue(
            transaction,
            EnqueueInput(
                owner="turns",
                kind=kind,
                source_ref={"chunkId": chunk_id},
                derivations=[
                    EnqueueDerivationTarget(
                        subject_kind="chunk",
                        subject_id=chunk_id,
                        derivation_type=kind,
                    )
                ],
            ),
        )
        for kind in ("chunk_summary_detailed", "chunk_summary_brief")
    ]


# The chunk structure for compact selection: every chunk in chunk order with
# its raw membership in member order. Membership is NOT filtered by turn
# liveness — references are returned as stored so the consumer can run its
# own referential check (a member pointing at no turn row at all is damage;
# a member pointing at a tombstoned turn is not). This is why it does not
# reuse read_chunk_rows, whose live-turn join would hide both cases.
@dataclass(frozen=True, slots=True)
class ChunkStructureRow:
    chunk_id: str
    chunk_order: int
    status: Literal["open", "closed"]
    member_turn_ids: list[str]


def read_chunk_structure(db: Database) -> list[ChunkStructureRow]:
    chunk_rows = db.prepare(_SQL_SELECT_CHUNK_STRUCTURE).all()
    member_rows = db.prepare(_SQL_SELECT_CHUNK_STRUCTURE_MEMBERS).all()
    members_by_chunk: dict[str, list[str]] = {}
    for row in member_rows:
        chunk_id = str(row["chunk_id"])
        members = members_by_chunk.get(chunk_id)
        if members is None:
            members = []
            members_by_chunk[chunk_id] = members
        members.append(str(row["turn_id"]))
    return [
        ChunkStructureRow(
            chunk_id=str(row["chunk_id"]),
            chunk_order=int(row["chunk_order"]),
            status=str(row["status"]),  # type: ignore[arg-type]
            member_turn_ids=list(members_by_chunk.get(str(row["chunk_id"]), [])),
        )
        for row in chunk_rows
    ]


# Placement read-back for the turns surface: chunk_id + member_idx by turn, one
# query, stored values only.
@dataclass(frozen=True, slots=True)
class TurnPlacement:
    chunk_id: str
    member_idx: int


def read_placements(db: Database) -> dict[str, TurnPlacement]:
    rows = db.prepare(_SQL_SELECT_PLACEMENTS).all()
    by_turn: dict[str, TurnPlacement] = {}
    for row in rows:
        by_turn[str(row["turn_id"])] = TurnPlacement(
            chunk_id=str(row["chunk_id"]),
            member_idx=int(row["member_idx"]),
        )
    return by_turn
